File: healthtrack/test_enhanced_slot_generation.dart
import 'dart:io';
import 'dart:convert';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;

void main() {
  group('Enhanced Slot Generation Tests', () {
    final String baseUrl = Platform.environment['BASE_URL'] ?? 'http://localhost:3000';
    
    // Test service ID (you may need to adjust this based on your test database)
    const testServiceId = 1;
    

    test('Single slot creation with valid parameters', () async {
      final tomorrow = DateTime.now().add(Duration(days: 1));
      final dateString = '${tomorrow.year}-${tomorrow.month.toString().padLeft(2, '0')}-${tomorrow.day.toString().padLeft(2, '0')}';
      
      final response = await http.post(
        Uri.parse('$baseUrl/appointment-slots'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'service_id': testServiceId,
          'appointment_date': dateString,
          'start_time': '09:00:00',
          'end_time': '09:30:00',
          'slot_duration_minutes': 30,
          'max_patients': 10,
        }),
      );
      
      expect(response.statusCode, 201);
      final data = json.decode(response.body);
      expect(data['success'], true);
      expect(data['data'], isNotNull);
    });

    test('Bulk slot generation with valid parameters', () async {
      final tomorrow = DateTime.now().add(Duration(days: 1));
      final dateString = '${tomorrow.year}-${tomorrow.month.toString().padLeft(2, '0')}-${tomorrow.day.toString().padLeft(2, '0')}';
      
      final response = await http.post(
        Uri.parse('$baseUrl/appointment-slots'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'service_id': testServiceId,
          'appointment_date': dateString,
          'start_time': '09:00:00',
          'end_time': '17:00:00',
          'slot_duration_minutes': 30,
          'max_patients': 10,
          'generate_slots': true,
        }),
      );
      
      expect(response.statusCode, 201);
      final data = json.decode(response.body);
      expect(data['success'], true);
      expect(data['data'], isList);
      expect(data['data'].length, greaterThan(0));
    });

    test('Slot creation with timeout handling', () async {
      final tomorrow = DateTime.now().add(Duration(days: 1));
      final dateString = '${tomorrow.year}-${tomorrow.month.toString().padLeft(2, '0')}-${tomorrow.day.toString().padLeft(2, '0')}';
      
      // This test verifies that the server handles timeout errors gracefully
      // We're not actually causing a timeout, but verifying the error structure
      try {
        final response = await http.post(
          Uri.parse('$baseUrl/appointment-slots'),
          headers: {'Content-Type': 'application/json'},
          body: json.encode({
            'service_id': testServiceId,
            'appointment_date': dateString,
            'start_time': 'invalid_time',
            'end_time': 'also_invalid',
          }),
        );
        
        // Expect validation error, not timeout
        expect(response.statusCode, 400);
      } catch (e) {
        // If we get a timeout exception, verify it's handled properly
        expect(e.toString(), contains('timed out'));
      }
    });

    test('Duplicate slot prevention', () async {
      final tomorrow = DateTime.now().add(Duration(days: 1));
      final dateString = '${tomorrow.year}-${tomorrow.month.toString().padLeft(2, '0')}-${tomorrow.day.toString().padLeft(2, '0')}';
      
      // First, create a slot
      await http.post(
        Uri.parse('$baseUrl/appointment-slots'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'service_id': testServiceId,
          'appointment_date': dateString,
          'start_time': '10:00:00',
          'end_time': '10:30:00',
          'slot_duration_minutes': 30,
          'max_patients': 10,
        }),
      );
      
      // Try to create a duplicate slot
      final response = await http.post(
        Uri.parse('$baseUrl/appointment-slots'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'service_id': testServiceId,
          'appointment_date': dateString,
          'start_time': '10:00:00',
          'end_time': '10:30:00',
          'slot_duration_minutes': 30,
          'max_patients': 10,
        }),
      );
      
      // Should get a conflict error
      expect(response.statusCode, 400);
      final data = json.decode(response.body);
      expect(data['success'], false);
      expect(data['message'], contains('already exist'));
    });

    test('Past date rejection', () async {
      final yesterday = DateTime.now().subtract(Duration(days: 1));
      final dateString = '${yesterday.year}-${yesterday.month.toString().padLeft(2, '0')}-${yesterday.day.toString().padLeft(2, '0')}';
      
      final response = await http.post(
        Uri.parse('$baseUrl/appointment-slots'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'service_id': testServiceId,
          'appointment_date': dateString,
          'start_time': '09:00:00',
          'end_time': '09:30:00',
        }),
      );
      
      expect(response.statusCode, 400);
      final data = json.decode(response.body);
      expect(data['success'], false);
      expect(data['message'], contains('past dates'));
    });

    test('Invalid time format rejection', () async {
      final tomorrow = DateTime.now().add(Duration(days: 1));
      final dateString = '${tomorrow.year}-${tomorrow.month.toString().padLeft(2, '0')}-${tomorrow.day.toString().padLeft(2, '0')}';
      
      final response = await http.post(
        Uri.parse('$baseUrl/appointment-slots'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'service_id': testServiceId,
          'appointment_date': dateString,
          'start_time': 'invalid_format',
          'end_time': 'also_invalid',
        }),
      );
      
      expect(response.statusCode, 400);
      final data = json.decode(response.body);
      expect(data['success'], false);
      expect(data['message'], contains('Invalid time format'));
    });
  });
}

File: healthtrack/test_enhanced_slot_generation.js
/**
 * Enhanced Slot Generation Test Script
 * Tests the enhanced appointment slot generation functionality
 */

const http = require('http');
const https = require('https');

// Configuration
const BASE_URL = process.env.BASE_URL || 'http://localhost:3000';
const TEST_SERVICE_ID = 1; // Adjust based on your test database

// Utility function to make HTTP requests
function makeRequest(options, postData = null) {
  return new Promise((resolve, reject) => {
    const lib = BASE_URL.startsWith('https') ? https : http;
    
    const req = lib.request(options, (res) => {
      let data = '';
      
      res.on('data', (chunk) => {
        data += chunk;
      });
      
      res.on('end', () => {
        try {
          resolve({
            statusCode: res.statusCode,
            headers: res.headers,
            body: JSON.parse(data)
          });
        } catch (e) {
          resolve({
            statusCode: res.statusCode,
            headers: res.headers,
            body: data
          });
        }
      });
    });
    
    req.on('error', (e) => {
      reject(e);
    });
    
    if (postData) {
      req.write(postData);
    }
    
    req.end();
  });
}

// Test functions
async function testSingleSlotCreation() {
  console.log('🧪 Test: Single slot creation with valid parameters');
  
  const tomorrow = new Date(Date.now() + 86400000);
  const dateString = tomorrow.toISOString().split('T')[0];
  
  const postData = JSON.stringify({
    service_id: TEST_SERVICE_ID,
    appointment_date: dateString,
    start_time: '09:00:00',
    end_time: '09:30:00',
    slot_duration_minutes: 30,
    max_patients: 10
  });
  
  const options = {
    hostname: new URL(BASE_URL).hostname,
    port: new URL(BASE_URL).port || (BASE_URL.startsWith('https') ? 443 : 80),
    path: '/appointment-slots',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(postData)
    }
  };
  
  try {
    const response = await makeRequest(options, postData);
    
    if (response.statusCode === 201) {
      console.log('   ✅ PASS: Single slot created successfully');
      return true;
    } else {
      console.log(`   ❌ FAIL: Expected status 201, got ${response.statusCode}`);
      console.log(`   Response: ${JSON.stringify(response.body)}`);
      return false;
    }
  } catch (error) {
    console.log(`   ❌ FAIL: Request failed with error: ${error.message}`);
    return false;
  }
}

async function testBulkSlotGeneration() {
  console.log('🧪 Test: Bulk slot generation with valid parameters');
  
  const tomorrow = new Date(Date.now() + 86400000);
  const dateString = tomorrow.toISOString().split('T')[0];
  
  const postData = JSON.stringify({
    service_id: TEST_SERVICE_ID,
    appointment_date: dateString,
    start_time: '09:00:00',
    end_time: '17:00:00',
    slot_duration_minutes: 30,
    max_patients: 10,
    generate_slots: true
  });
  
  const options = {
    hostname: new URL(BASE_URL).hostname,
    port: new URL(BASE_URL).port || (BASE_URL.startsWith('https') ? 443 : 80),
    path: '/appointment-slots',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(postData)
    }
  };
  
  try {
    const response = await makeRequest(options, postData);
    
    if (response.statusCode === 201) {
      if (Array.isArray(response.body.data) && response.body.data.length > 0) {
        console.log(`   ✅ PASS: Bulk slots created successfully (${response.body.data.length} slots)`);
        return true;
      } else {
        console.log('   ❌ FAIL: Expected array of slots in response data');
        return false;
      }
    } else {
      console.log(`   ❌ FAIL: Expected status 201, got ${response.statusCode}`);
      console.log(`   Response: ${JSON.stringify(response.body)}`);
      return false;
    }
  } catch (error) {
    console.log(`   ❌ FAIL: Request failed with error: ${error.message}`);
    return false;
  }
}

async function testDuplicateSlotPrevention() {
  console.log('🧪 Test: Duplicate slot prevention');
  
  const tomorrow = new Date(Date.now() + 86400000);
  const dateString = tomorrow.toISOString().split('T')[0];
  
  // First, create a slot
  const postData1 = JSON.stringify({
    service_id: TEST_SERVICE_ID,
    appointment_date: dateString,
    start_time: '10:00:00',
    end_time: '10:30:00',
    slot_duration_minutes: 30,
    max_patients: 10
  });
  
  const options1 = {
    hostname: new URL(BASE_URL).hostname,
    port: new URL(BASE_URL).port || (BASE_URL.startsWith('https') ? 443 : 80),
    path: '/appointment-slots',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(postData1)
    }
  };
  
  try {
    // Create first slot
    await makeRequest(options1, postData1);
    
    // Try to create duplicate slot
    const response = await makeRequest(options1, postData1);
    
    if (response.statusCode === 400) {
      if (response.body.message && response.body.message.includes('already exist')) {
        console.log('   ✅ PASS: Duplicate slot properly rejected');
        return true;
      } else {
        console.log('   ❌ FAIL: Expected duplicate slot error message');
        return false;
      }
    } else {
      console.log(`   ❌ FAIL: Expected status 400, got ${response.statusCode}`);
      return false;
    }
  } catch (error) {
    console.log(`   ❌ FAIL: Request failed with error: ${error.message}`);
    return false;
  }
}

async function testPastDateRejection() {
  console.log('🧪 Test: Past date rejection');
  
  const yesterday = new Date(Date.now() - 86400000);
  const dateString = yesterday.toISOString().split('T')[0];
  
  const postData = JSON.stringify({
    service_id: TEST_SERVICE_ID,
    appointment_date: dateString,
    start_time: '09:00:00',
    end_time: '09:30:00'
  });
  
  const options = {
    hostname: new URL(BASE_URL).hostname,
    port: new URL(BASE_URL).port || (BASE_URL.startsWith('https') ? 443 : 80),
    path: '/appointment-slots',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(postData)
    }
  };
  
  try {
    const response = await makeRequest(options, postData);
    
    if (response.statusCode === 400) {
      if (response.body.message && response.body.message.includes('past dates')) {
        console.log('   ✅ PASS: Past date properly rejected');
        return true;
      } else {
        console.log('   ❌ FAIL: Expected past date error message');
        return false;
      }
    } else {
      console.log(`   ❌ FAIL: Expected status 400, got ${response.statusCode}`);
      return false;
    }
  } catch (error) {
    console.log(`   ❌ FAIL: Request failed with error: ${error.message}`);
    return false;
  }
}

async function testInvalidTimeFormatRejection() {
  console.log('🧪 Test: Invalid time format rejection');
  
  const tomorrow = new Date(Date.now() + 86400000);
  const dateString = tomorrow.toISOString().split('T')[0];
  
  const postData = JSON.stringify({
    service_id: TEST_SERVICE_ID,
    appointment_date: dateString,
    start_time: 'invalid_format',
    end_time: 'also_invalid'
  });
  
  const options = {
    hostname: new URL(BASE_URL).hostname,
    port: new URL(BASE_URL).port || (BASE_URL.startsWith('https') ? 443 : 80),
    path: '/appointment-slots',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(postData)
    }
  };
  
  try {
    const response = await makeRequest(options, postData);
    
    if (response.statusCode === 400) {
      if (response.body.message && response.body.message.includes('Invalid time format')) {
        console.log('   ✅ PASS: Invalid time format properly rejected');
        return true;
      } else {
        console.log('   ❌ FAIL: Expected invalid time format error message');
        return false;
      }
    } else {
      console.log(`   ❌ FAIL: Expected status 400, got ${response.statusCode}`);
      return false;
    }
  } catch (error) {
    console.log(`   ❌ FAIL: Request failed with error: ${error.message}`);
    return false;
  }
}

// Main test runner
async function runTests() {
  console.log('🚀 Starting Enhanced Slot Generation Tests\n');
  
  let passedTests = 0;
  const totalTests = 5;
  
  // Run all tests
  if (await testSingleSlotCreation()) passedTests++;
  console.log('');
  
  if (await testBulkSlotGeneration()) passedTests++;
  console.log('');
  
  if (await testDuplicateSlotPrevention()) passedTests++;
  console.log('');
  
  if (await testPastDateRejection()) passedTests++;
  console.log('');
  
  if (await testInvalidTimeFormatRejection()) passedTests++;
  console.log('');
  
  // Summary
  console.log(`\n🏁 Test Results: ${passedTests}/${totalTests} tests passed`);
  
  if (passedTests === totalTests) {
    console.log('🎉 All tests passed!');
    process.exit(0);
  } else {
    console.log('❌ Some tests failed.');
    process.exit(1);
  }
}

// Run the tests
runTests().catch(error => {
  console.error('Test suite failed with unhandled error:', error);
  process.exit(1);
});